### ISSUE TITLE
`io.ErrShortBuffer` not handled in datachannel readLoop()

### ISSUE THREAD

**Issue by cohosh** [pion/webrtc#718]
### Your environment.
  - Version: Release
  - Browser: N/A (using library in Go Application)
  - Other Information - see https://trac.torproject.org/projects/tor/ticket/28942#comment:14

### What did you do?
We used the pion/webrtc library for one end of communication in our Snowflake anti-censorship system. The other end uses keroserene/go-webrtc.

### What did you expect?
I expected to receive all of the data sent by the other end of the webrtc connection.

### What happened?
 We found that large chunks of data were being dropped despite the usage of a reliable channel. Upon further investigation, we found the reassembly queue returning an `io.ErrShortBuffer` error.

**Comment by Sean-Der** [comment]
@enobufs @backkem 

I would really like to see Snowflake use Pion. Me and @backkem talked about it being a big achievement if we could make Pion good enough for them almost a year back.

Anyone I can bribe one of you two to help out with this :) I am working on triaging issues down and DTLS PSK. If no one has bandwidth I should be able to help this weekend.

-----
Some thoughts/questions though.

* Is there any reason we can't/shouldn't grow the buffer if it was too short before?
* Can we make this configurable via `SettingEngine`, still have a sensible default but document this behavior.

The detached data channels is great, but would be nice if we can fix the 'sharp edge' here. Never want to surprise users.

thanks :)

**Comment by enobufs** [comment]
I have just reviewed @cohosh's PR at pion/sctp#51 and it looks good for what he needs. The PR does:
(1) Return io.ErrSHortBuffer properly (<- this is definitely a bug to fix)
(2) Leave the message in the reassemblyQue for read retry.

(2) is what I am not sure about and looking for WebRTC spec for it. My initial thoughts was, the UDP socket for instance, if the provided buffer is too small, the packet will be lost (as far as I remember).  From SCTP's perspective, allowing large message would cause other channels not being able to send during the transfer of the large message. To overcome it, I learned there was an SCTP extension, "Stream Schedulers and User Message Interleaving" (RFC 8260) to solve the blocking problem. Chrome and Firefox apparently implement it (not 100% sure).

**Comment by enobufs** [comment]
Regarding the growing the buffer size, my first reaction to it is, growing the size indefinitely is not a good idea, but maybe we could at least increase the static buffer size to 32KB or 64KB?

**Comment by cohosh** [comment]
Yeah, I can see how both of the design decisions of (2) leaving data in the queue on error, and continuously growing the buffer size can each set ourselves up for a situation in which we're in an infinite loop. I'll see if I can piece together what the Chrome implementation does. It might be worth separating this into a two PRs to handle (1) and (2) separately.

**Comment by cohosh** [comment]
Hi,
I did some investigating, including reading the article linked by @enobufs: https://lgrahl.de/articles/demystifying-webrtc-dc-size-limit.html

I turns out that we were experiencing problems even with using the pion/webrtc implementation on the client side of our system. This is due to the fact that pion/webrtc implements a maximum send size of 65536 (math.MaxUint16) in WriteSCTP https://github.com/pion/sctp/blob/master/stream.go#L146

The dataChannelBufferSize is set to significantly less than this. Increasing the buffer size will at least solve the problem for messages sent by the same implementation. As far as using other implementations at different peers, it looks like logging the error and closing the data channel is a better solution than either failing silently as before, or logging and continuing to read data. At least for our purposes, we rely on reliable and ordered data channels and having some of our data simply go missing causes significant problems.

I updated both pull requests to remove the changes the reassembly_queue but keep the return of `io.ErrShortBuffer`: https://github.com/pion/sctp/pull/51
And also the PR about handling the error to increase the buffer size to 64KiB and close the datachannel on error: https://github.com/pion/webrtc/pull/719
